src/main/java/com/techsophy/tsf/account/service/impl/ACLServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+import com.techsophy.tsf.account.dto.ACLSchema;
+import com.techsophy.tsf.account.dto.ACLValidate;
+import com.techsophy.tsf.account.dto.ACLDecision;
+import com.techsophy.tsf.account.entity.ACLDefinition;
+import com.techsophy.tsf.account.exception.EntityNotFoundByIdException;
+import com.techsophy.tsf.account.exception.UserDetailsIdNotFoundException;
+import com.techsophy.tsf.account.repository.ACLRepository;
+import com.techsophy.tsf.account.service.ACLService;
+import com.techsophy.tsf.account.utils.TokenUtils;
+import com.techsophy.tsf.account.utils.UserDetails;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+import java.math.BigInteger;
+import java.nio.file.AccessDeniedException;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Objects;
+
+import static com.techsophy.tsf.account.constants.AccountConstants.*;
+import static com.techsophy.tsf.account.constants.ErrorConstants.ACL_NOT_FOUND_WITH_GIVEN_ID;
+import static com.techsophy.tsf.account.constants.ErrorConstants.LOGGED_IN_USER_ID_NOT_FOUND;
+import static org.apache.commons.lang3.ObjectUtils.isEmpty;
+
+@Slf4j
+@Service
+@AllArgsConstructor(onConstructor_ = {@Autowired})
+public class ACLServiceImpl implements ACLService
+{
+    private final GlobalMessageSource globalMessageSource;
+    private final IdGeneratorImpl idGenerator;
+    private final ObjectMapper objectMapper;
+    private final UserDetails userDetails;
+    private final ACLRepository aclRepository;
+    private final TokenUtils tokenUtils;
+
+    @Override
+    public ACLSchema saveACL(ACLSchema aclSchema) throws JsonProcessingException
+    {
+        Map<String,Object> loggedInUserDetails =userDetails.getUserDetails().get(0);
+        if (isEmpty(String.valueOf(loggedInUserDetails.get(ID))))
+        {
+            throw new UserDetailsIdNotFoundException(LOGGED_IN_USER_ID_NOT_FOUND,globalMessageSource.get(LOGGED_IN_USER_ID_NOT_FOUND,loggedInUserDetails.get(ID).toString()));
+        }
+        BigInteger loggedInUserId = BigInteger.valueOf(Long.parseLong(loggedInUserDetails.get(ID).toString()));
+        ACLDefinition aclDefinition=this.objectMapper.convertValue(aclSchema,ACLDefinition.class);
+        aclDefinition.setUpdatedById(loggedInUserId);
+        if(aclSchema.getId()!=null)
+        {
+            try {
+                return aclRepository.updateACLDefinition(aclSchema,loggedInUserId);
+            }
+            catch (Exception e)
+            {
+                throw new DuplicateKeyException("Duplicate names are not allowed for ACL");
+            }
+        }
+        BigInteger id=idGenerator.nextId();
+        aclDefinition.setId(id);
+        aclDefinition.setCreatedById(loggedInUserId);
```

Comment:
these should be replaced by mongo auditing

Reply:
yes replacing with mongoAudit